crates/moveos/src/vm/tx_argument_resolver.rs

```diff
@@ -0,0 +1,199 @@
+    tx_context::TxContext,
+};
+use move_binary_format::errors::PartialVMError;
+use move_core_types::{move_resource::MoveStructType, value::MoveValue, vm_status::StatusCode};
+use move_vm_runtime::session::LoadedFunctionInstantiation;
+use move_vm_types::loaded_data::runtime_types::{StructType, Type};
+use std::sync::Arc;
+
+use super::{move_vm_ext::SessionExt, MoveResolverExt};
+
+/// Transaction Argument Resolver will implemented by the Move Extension
+/// to auto fill transaction argument or do type conversion.
+// TODO: Try to push to Move upstream if possible.
+pub trait TxArgumentResolver {
+    fn resolve_argument<S>(
+        &self,
+        session: &SessionExt<S>,
+        func: &LoadedFunctionInstantiation,
+        args: Vec<Vec<u8>>,
+    ) -> Result<Vec<Vec<u8>>, PartialVMError>
+    where
+        S: MoveResolverExt;
+}
+
+impl TxArgumentResolver for TxContext {
+    fn resolve_argument<S>(
+        &self,
+        session: &SessionExt<S>,
+        func: &LoadedFunctionInstantiation,
+        mut args: Vec<Vec<u8>>,
+    ) -> Result<Vec<Vec<u8>>, PartialVMError>
+    where
+        S: MoveResolverExt,
+    {
+        let has_signer = func
+            .parameters
+            .iter()
+            .position(is_signer)
+            .map(|pos| {
+                if pos != 0 {
+                    Err(
+                        PartialVMError::new(StatusCode::NUMBER_OF_SIGNER_ARGUMENTS_MISMATCH)
+                            .with_message(format!(
+                                "Expected signer arg is this first arg, but got it at {}",
+                                pos + 1
+                            )),
+                    )
+                } else {
+                    Ok(true)
+                }
+            })
+            .unwrap_or(Ok(false))?;
+
+        if has_signer {
+            let signer = MoveValue::Signer(self.sender());
+            args.insert(
+                0,
+                signer
+                    .simple_serialize()
+                    .expect("serialize signer should success"),
+            );
+        }
+
+        let has_tx_context = func
+            .parameters
+            .iter()
+            .position(|i| {
+                is_object(session, i)
+                    .map(|t| is_tx_context(&t))
+                    .unwrap_or(false)
+            })
+            .map(|pos| {
+                if pos != 0 {
+                    Err(
+                        PartialVMError::new(StatusCode::NUMBER_OF_SIGNER_ARGUMENTS_MISMATCH)
+                            .with_message(format!(
+                                "Expected TxContext arg is this first arg, but got it at {}",
+                                pos + 1
+                            )),
+                    )
+                } else {
+                    Ok(true)
+                }
+            })
+            .unwrap_or(Ok(false))?;
+
+        if has_tx_context {
+            args.insert(0, self.to_vec());
+        }
```

Comment:
Why are has_signer and  has_tx_context  both insert to 0 index position in args?

Reply:
`signer` and `TxContext` only can be the first argument of the transaction.